src/azul/queues.py

```diff
@@ -149,66 +150,104 @@ def get_queues(self, queue_names: Iterable[str]) -> Mapping[str, Queue]:
+            if max_timeout is not None:
+                logger.info('Maximum wait time set to %s', max_timeout)
+                timeout = min(timeout, max_timeout)
+            elif min_timeout is not None:
+                logger.info('Minimum wait time set to %s', max_timeout)
+                timeout = max(min_timeout, timeout)
+            return timeout
+
+        if max_timeout is not None and min_timeout is not None:
+            logger.error('The parameters `max_timeout` and `min_timeout` '
+                         'should not be specified simultaneously.')
+            raise Exception
+
         sleep_time = 5
         deque_size = 10 if empty else 1
         queues = self.get_queues(config.work_queue_names)
         queue_size_history = deque(maxlen=deque_size)
         wait_start_time = time.time()
+        compute_timeout = None
 
         if not empty:
             timeout = 2 * 60
-        elif num_bundles is None:
-            timeout = 60 * 60
         else:
-            # calculate timeout for queues to empty with a given number of bundles
-            time_to_clear_deque = deque_size * sleep_time
-            time_processing_bundles = num_bundles * 5  # small time per bundle
-            timeout = max(time_processing_bundles + time_to_clear_deque, 60)  # at least 1 min
-            timeout = min(timeout, 60 * 60)  # at most 60 min
+            if num_bundles is None:
+                queue_message_count = self.count_messages(queues)
+                num_bundles = queue_message_count[config.notify_queue_name]
+            compute_timeout = _compute_timeout
+            timeout = compute_timeout(num_bundles)
+            logger.info('Wait time was derived from the available bundles.')
+        iterations_in_one_minute = 60 / sleep_time
 
         logger.info('Waiting up to %s seconds for %s queues to %s ...',
                     timeout, len(queues), 'empty' if empty else 'not be empty')
         while True:
-            total_message_count = self.count_messages(queues)
+            queue_message_count = self.count_messages(queues)
+            total_message_count = queue_message_count['total_message_count']
             logger.info('Counting %i total message(s) in %i queue(s).', total_message_count, len(queues))
             queue_wait_time_elapsed = (time.time() - wait_start_time)
             queue_size_history.append(total_message_count)
             cumulative_queue_size = sum(queue_size_history)
+            if compute_timeout is not None:
+                num_b = queue_message_count[config.notify_queue_name]
+                updated_timeout = compute_timeout(num_b)
+                if updated_timeout > timeout:
+                    timeout = updated_timeout
+                    logger.info('Wait time updated to %i seconds.', timeout)
+                else:
+                    iterations_in_one_minute -= 1
+                    if iterations_in_one_minute <= 0:
```

Comment:
Is there a better name for this variable? `sleep_time` doesn't change dynamically so when `iterations_in_one_minute` is decremented it to no longer represents the actual iterations in one minute.